src/DataGrid.tsx

```diff
@@ -320,25 +327,32 @@ export function DataGrid<R, SR = unknown, K extends Key = Key>(props: DataGridPr
    */
   const [scrollTop, setScrollTop] = useState(0);
   const [scrollLeft, setScrollLeft] = useState(0);
-  const [resizedColumnWidths, setResizedColumnWidths] = useState(
-    (): ReadonlyMap<string, number> => new Map()
-  );
-  const [measuredColumnWidths, setMeasuredColumnWidths] = useState(
-    (): ReadonlyMap<string, number> => new Map()
+  const [columnWidthsInternal, setColumnWidthsInternal] = useState(
+    (): ColumnWidths => columnWidthsRaw ?? new Map()
   );
+  const [isColumnResizing, setColumnResizing] = useState(false);
   const [isDragging, setDragging] = useState(false);
   const [draggedOverRowIdx, setOverRowIdx] = useState<number | undefined>(undefined);
   const [scrollToPosition, setScrollToPosition] = useState<PartialPosition | null>(null);
   const [shouldFocusCell, setShouldFocusCell] = useState(false);
   const [previousRowIdx, setPreviousRowIdx] = useState(-1);
 
+  const isColumnWidthsControlled =
+    columnWidthsRaw != null && onColumnWidthsChangeRaw != null && !isColumnResizing;
+  const columnWidths = isColumnWidthsControlled ? columnWidthsRaw : columnWidthsInternal;
```

Comment:
Internal state serves two purposes 
1. It is used when the props are not provided
2. It is used for setting temporary until the resize operation is complete